superbench/analyzer/data_diagnosis.py

```diff
@@ -0,0 +1,359 @@
+            # check if metric pass the rule
+            else:
+                data = data_row[metric]
+                rule = self._sb_baseline[metric]['rules']
+                rule_op = RuleOp.get_rule_func(RuleType(rule['name']))
+                (pass_rule, processed) = rule_op(data, baseline, rule)
+                summary_data_row[metric] = processed
+            # label the node as issued one
+            if not pass_rule:
+                # use return code to identify 'miss test'
+                if 'return_code' in metric:
+                    category_details.append(benchmark + '_miss')
+                    categories.add('MissTest')
+                    issue_label = True
+                else:
+                    category_details.append(metric)
+                    categories.add(benchmark)
+                    if benchmark not in cnn_benchmarks:
+                        issue_label = True
+                    # for cnn models, only 2 model metric issued, label as issue
+                    else:
+                        model_label_num += 1
+                        if model_label_num >= 2:
+                            issue_label = True
+
+        if issue_label:
+            # Add category information
+            general_cat_str = ','.join(categories)
+            details_cat_str = ','.join(category_details)
+            hw_issue_flag = self.hw_issue(categories)
+            details_row = [hw_issue_flag, len(category_details), general_cat_str, details_cat_str]
+            return (details_row, summary_data_row)
+
+        return (None, None)
+
+    def rule_based_diagnosis(self, rule_file):
+        """Rule-based data diagnosis for multi nodes' raw data.
+
+        Use the rules defined in rule_file to diagnose the raw data of each node,
+        if the node violate any rule, label as issued node and save
+        the 'Hw Issues', '# of Issues', 'Category', 'Issue Details' and processed data of issued node.
+
+        Args:
+            rule_file (str): The path of baseline yaml file
+
+        Returns:
+            DataFrame: issued nodes's detailed information
+        """
+        if len(self._raw_data_df) == 0:
+            return False
+        if not self._get_criteria(rule_file):
+            return False
+        summary_columns = ['Hw Issues', '# of Issues', 'Category', 'Issue Details']
+        data_not_accept_df = pd.DataFrame(columns=summary_columns)
+        summary_details_df = pd.DataFrame()
+
+        for node in self._raw_data_df.index:
+            (details_row, summary_data_row) = self.single_node_diagnosis(node)
+            if details_row:
+                data_not_accept_df.loc[node] = details_row
+                summary_details_df = summary_details_df.append(summary_data_row)
+
+        data_not_accept_df = data_not_accept_df.join(summary_details_df)
+        data_not_accept_df = data_not_accept_df.sort_values(by=summary_columns, ascending=False)
+        return data_not_accept_df
+
```

Comment:
Do we need to provide a separate class for excel output?

Reply:
separate a file for this